tests/test_types.py

```diff
@@ -106,10 +106,37 @@ def test_parse_type_no_basic_types():
     nose.tools.assert_true(byte.size, 8)
     nose.tools.assert_false(byte.signed)
 
+def test_self_referential_struct_or_union():
+    struct_llist = angr.types.parse_type('struct llist { int data; struct llist *next; }')
+    next_struct_llist = struct_llist.fields['next'].pts_to
+    nose.tools.assert_true(len(next_struct_llist.fields) == 2)
+    nose.tools.assert_is_instance(next_struct_llist.fields['data'], SimTypeInt)
+    nose.tools.assert_is_instance(next_struct_llist.fields['next'], SimTypePointer)
+
+    union_heap = angr.types.parse_type('union heap { int data; union heap *forward; }')
+    forward_union_heap = union_heap.members['forward'].pts_to
+    nose.tools.assert_true(len(forward_union_heap.members) == 2)
+    nose.tools.assert_is_instance(forward_union_heap.members['data'], SimTypeInt)
+    nose.tools.assert_is_instance(forward_union_heap.members['forward'], SimTypePointer)
+
+def test_avoid_modify_stored_union_struct_while_parsing():
+    union_a = angr.types.parse_type('union a')
+    angr.types.register_types(union_a)
+    angr.types.parse_type('union a { int x; }')
+    nose.tools.assert_true(len(union_a.members) == 0)
+
+def test_avoid_return_stored_union_struct_while_parsing():
+    union_a = angr.types.parse_type('union a { int x; }')
+    angr.types.register_types(union_a)
+    parsed_union_a = angr.types.parse_type('union a')
+    nose.tools.assert_false(union_a is parsed_union_a)
```

Comment:
I think this is incorrect!! Once you've registered the type, all future references to it are supposed to be that type you registered!

also use the right nose assertion

Reply:
You're right! Besides this, I have thought better and avoiding modify the stored type does not allow to define structs/unions referecing each other, what is a expected behavior.

```python
>>> angr.types.register_types(angr.types.parse_type('struct a'))
>>> angr.types.register_types(angr.types.parse_type('struct b'))
>>> struct_a = angr.types.parse_type('struct a { struct b *b_ptr; }')
>>> struct_b = angr.types.parse_type('struct b { struct a *a_ptr; }')
>>> struct_a.fields
OrderedDict([('b_ptr', struct b*)])
>>> struct_b.fields
OrderedDict([('a_ptr', struct a*)])
```

If it is OK to you, I will remove all the changes from the last commit and add the structs referecing each other test.

Reply:
Sorry about the mess!